### PR TITLE
fixing matrix derivative bugs

### PR DESCRIPTION
**Pull request by Upabjojr**
#### Release Notes

* matrices
  * fixing some common bugs with derivatives of matrix expressions.

### CODE REVIEW

File: sympy/matrices/expressions/matexpr.py [hunk 1]
```diff
@@ -976,6 +989,12 @@ def transpose(self):
     def matrix_form(self):
         if self.first != 1 and self.higher != 1:
             raise ValueError("higher dimensional array cannot be represented")
+        # Remove one-dimensional identity matrices:
+        # (this is needed by `a.diff(a)` where `a` is a vector)
+        if self.first == Identity(1):
+            return self.second.T
```
Comment (oscarbenjamin) [inline sympy/matrices/expressions/matexpr.py]: There is no test for this branch

File: sympy/matrices/expressions/tests/test_derivatives.py [hunk 2]
```diff
@@ -246,9 +250,23 @@ def test_derivatives_of_complicated_matrix_expr():
 
 def test_mixed_deriv_mixed_expressions():
 
+    expr = 3*Trace(A)
+    assert expr.diff(A) == 3*Identity(k)
+
+    expr = k
+    deriv = expr.diff(A)
+    assert isinstance(deriv, ZeroMatrix)
```
Comment (oscarbenjamin) [inline sympy/matrices/expressions/tests/test_derivatives.py]: Is this not tested by the line below?

**Comment by oscarbenjamin** [comment]
I haven't been through all of them but from the examples I've tried the new tests already pass on master. Not saying those tests shouldn't be added but can you explain in the PR description what has actually been fixed/changed here?

File: sympy/matrices/expressions/matexpr.py (hunk 1 above)
Comment (Upabjojr) [reply sympy/matrices/expressions/matexpr.py]: Line 62 of the test file.

File: sympy/matrices/expressions/tests/test_derivatives.py (hunk 2 above)
Comment (Upabjojr) [reply sympy/matrices/expressions/tests/test_derivatives.py]: There are cases in which `==` returns `True` even if the objects do not belong to the same class.

**Comment by Upabjojr** [comment]
> I haven't been through all of them but from the examples I've tried the new tests already pass on master. Not saying those tests shouldn't be added but can you explain in the PR description what has actually been fixed/changed here?

Did you try out the correct commit on master? I have already merged these commits to the master branch.

File: sympy/matrices/expressions/matexpr.py (hunk 1 above)
Comment (oscarbenjamin) [reply sympy/matrices/expressions/matexpr.py]: If you tried his diff:
```diff
$ git diff
diff --git a/sympy/matrices/expressions/matexpr.py b/sympy/matrices/expressions/matexpr.py
index 5f54bb907f..af51dd8a29 100644
--- a/sympy/matrices/expressions/matexpr.py
+++ b/sympy/matrices/expressions/matexpr.py
@@ -992,7 +992,7 @@ def matrix_form(self):
         # Remove one-dimensional identity matrices:
         # (this is needed by `a.diff(a)` where `a` is a vector)
         if self.first == Identity(1):
-            return self.second.T
+            return 1/0
         if self.second == Identity(1):
             return self.first
         if self.first != 1:
```
and all tests in sympy.matrices still passed.

**Comment by oscarbenjamin** [comment]
> Did you try out the correct commit on master? I have already merged these commits to the master branch.

Okay I see what you mean. I tried them now with 1.4 since this is what you want to merge them into and I can see that the examples don't work on 1.4 without this PR.

**Pull request merged**